server/Arcadia.Assistant.Web/Download/DownloadActor.cs

```diff
@@ -0,0 +1,186 @@
+                    if (msg.Message != null)
+                    {
+                        this.logger.Warning(msg.Message);
+                    }
+
+                    break;
+
+                default:
+                    this.Unhandled(message);
+                    break;
+            }
+        }
+
+        private async Task DownloadApplicationBuilds()
+        {
+            var androidDownloadBuildTask = this.DownloadBuild(
+                this.downloadApplicationSettings.AndroidGetBuildsUrl,
+                this.downloadApplicationSettings.AndroidGetBuildDownloadLinkTemplateUrl);
+
+            var iosDownloadBuildTask = this.DownloadBuild(
+                this.downloadApplicationSettings.IosGetBuildsUrl,
+                this.downloadApplicationSettings.IosGetBuildDownloadLinkTemplateUrl);
+
+            await Task.WhenAll(androidDownloadBuildTask, iosDownloadBuildTask);
+        }
+
+        private async Task DownloadBuild(string getBuildsUrl, string getBuildDownloadLinkTemplateUrl)
+        {
+            var latestBuild = await this.GetLatestBuild(getBuildsUrl);
+            if (latestBuild == null)
+            {
+                this.logger.Warning($"No builds found for url {getBuildsUrl}");
+                return;
+            }
+
+            var buildDownloadModel = await this.GetBuildDownloadModel(getBuildDownloadLinkTemplateUrl, latestBuild);
+
+            using (var client = this.httpClientFactory.CreateClient())
+            {
+                var buildFileStream = await client.GetStreamAsync(buildDownloadModel.Uri);
+                this.SaveBuildFile(buildFileStream, buildDownloadModel);
+            }
+        }
+
+        private async Task<AppCenterBuildModel> GetLatestBuild(string getBuildsUrl)
+        {
+            using (var response = await this.SendAppCenterRequest(getBuildsUrl))
+            {
+                var contentString = await response.Content.ReadAsStringAsync();
+                var builds = this.DeserializeJson<IEnumerable<AppCenterBuildModel>>(contentString);
+
+                return builds
+                    .OrderByDescending(b => b.FinishTime)
+                    .FirstOrDefault();
+            }
+        }
+
+        private async Task<AppCenterBuildDownloadModel> GetBuildDownloadModel(string getBuildDownloadLinkTemplateUrl, AppCenterBuildModel build)
+        {
+            var getBuildUrl = getBuildDownloadLinkTemplateUrl.Replace("{buildId}", build.Id.ToString());
+            using (var response = await this.SendAppCenterRequest(getBuildUrl))
+            {
+                var contentString = await response.Content.ReadAsStringAsync();
+
+                var downloadModel = this.DeserializeJson<AppCenterBuildDownloadModel>(contentString);
+                downloadModel.BuildDate = build.FinishTime;
+
+                return downloadModel;
+            }
+        }
+
+        private void SaveBuildFile(Stream fileStream, AppCenterBuildDownloadModel buildDownloadModel)
```

Comment:
why is this sync method when it does IO operation

Reply:
something is wrong, it should not happen. what's the difference between this and await?

Reply:
Done.